File: api/addressbook.go
package api

import (
	"context"
	"github.com/iain17/decentralizer/pb"
)

//
// Address book
//
// Create or update a peer. Takes peer info, returns if it was a success.
func (s *Server) UpsertPeer(ctx context.Context, request *pb.RPCUpsertPeerRequest) (*pb.RPCUpsertPeerResponse, error) {
	_, err := s.App.UpsertPeer(request.Peer.PId, request.Peer.Details)
	if err == nil && request.Peer.PId == "self" {
		err = s.App.AdvertisePeerRecord()
	}
	return &pb.RPCUpsertPeerResponse{}, err
}

// Get peer ids. takes a key and value to filter the peers by details. If left empty this filter will not apply and all will be fetched.
func (s *Server) GetPeerIds(ctx context.Context, request *pb.RPCGetPeerIdsRequest) (*pb.RPCGetPeerIdsResponse, error) {
	var peers []*pb.Peer
	var err error
	if request.Key == "" && request.Value == "" {
		peers, err = s.App.GetPeers()
	} else {
		peers, err = s.App.GetPeersByDetails(request.Key, request.Value)
	}
	var peerIds []string
	for _, peer := range peers {
		peerIds = append(peerIds, peer.PId)
	}
	return &pb.RPCGetPeerIdsResponse{
		PeerIds: peerIds,
	}, err
}

// Get an individual peer. Takes either a peer id or decentralizer id and returns the peer info.
func (s *Server) GetPeer(ctx context.Context, request *pb.RPCGetPeerRequest) (*pb.RPCGetPeerResponse, error) {
	var peer *pb.Peer
	var err error
	if request.PId != "" {
		peer, err = s.App.FindByPeerId(request.PId)
	}
	if peer == nil && request.DnId != 0 {
		peer, err = s.App.FindByDecentralizedId(request.DnId)
	}
	return &pb.RPCGetPeerResponse{
		Peer: peer,
	}, err
}

File: sdk/libdn/libdn/Addressbook.cpp
#include "StdInc.h"

namespace libdn {
	LIBDN_API std::shared_ptr<Promise<bool>> LIBDN_CALL UpsertPeer(libdn::Peer& peer) {
		auto result = std::make_shared<Promise<bool>>([peer](auto promise) {
			// Data we are sending to the server.
			pb::RPCUpsertPeerRequest request;
			pb::Peer p = DNPeerToPBPeer(peer);
			request.set_allocated_peer(&p);

			// Container for the data we expect from the server.
			pb::RPCUpsertPeerResponse reply;

			auto ctx = context.client->getContext();
			grpc::Status status = context.client->stub_->UpsertPeer(ctx, request, &reply);

			if (!status.ok()) {
				promise->reject(fmt::format("[Could not upsert peer] {0}: {1}", status.error_code(), status.error_message().c_str()));
				return false;
			}
			return true;
		});
		return result;
	}

	LIBDN_API std::shared_ptr<Promise<int>> LIBDN_CALL GetNumPeers(const char * key, const char * value) {
		auto result = std::make_shared<Promise<int>>([key, value](auto promise) {
			//build request.
			pb::RPCGetPeerIdsRequest request;
			request.set_key(key);
			request.set_value(value);

			// Container for the data we expect from the server.
			pb::RPCGetPeerIdsResponse reply;

			auto ctx = context.client->getContext();
			grpc::Status status = context.client->stub_->GetPeerIds(ctx, request, &reply);

			if (status.ok()) {
				context.peers = reply.peerids();
				int size = context.peers.size();
				return size;
			} else {
				promise->reject(fmt::format("[Could not get peer ids] {0}: {1}", status.error_code(), status.error_message().c_str()));
			}

			return 0;
		});
		return result;
	}

	//If pId is set to "self" it will automatically resolve to the local peer id.
	//If both are given. Perference is given to pId.
	LIBDN_API std::shared_ptr<Promise<Peer>> LIBDN_CALL GetPeerById(DNID dId, PeerID& pId) {
		auto result = std::make_shared<Promise<Peer>>([dId, pId](auto promise) {
			//build request.
			pb::RPCGetPeerRequest request;
			request.set_dnid(dId);
			request.set_pid(pId);

			// Container for the data we expect from the server.
			pb::RPCGetPeerResponse reply;

			auto ctx = context.client->getContext();
			grpc::Status status = context.client->stub_->GetPeer(ctx, request, &reply);

			if (!status.ok()) {
				promise->reject(fmt::format("[Could not get peer] {0}: {1}", status.error_code(), status.error_message().c_str()));
			}
			return PBPeerToDNPeer(reply.peer());
		});
		return result;
	}

	LIBDN_API Peer* LIBDN_CALL GetSelf() {
		context.selfMutex.lock();
		Peer* self = &context.self;
		context.selfMutex.unlock();
		return self;
	}

	LIBDN_API std::shared_ptr<Peer> LIBDN_CALL GetPeerByIndex(int index) {
		if (index > context.peers.size() - 1) {
			return NULL;
		}
		std::string pid = context.peers.Get(index);
		auto req = GetPeerById(0, pid);
		req->fail([](std::string reason) {
			Log_Print(reason.c_str());
		});
		if (req->wait()) {
			return std::make_shared<Peer>(req->get());
		}
		return NULL;
	}

	LIBDN_API std::shared_ptr<PeerID> LIBDN_CALL ResolveDecentralizedId(DNID dId) {
		std::string empty = "";
		auto request = GetPeerById(dId, empty);
		request->fail([](const char* reason) {
			Log_Print("Could not resolve DecentralizedId: %s", reason);
		});
		if (request->wait()) {
			auto test = request->get().pId;
			return std::make_shared<PeerID>(test);
		}
		return std::make_shared<PeerID>("");
	}
}